In [16]:

import pandas as pd
import pprint
from collections import OrderedDict
from datetime import datetime


import time
import mido

fname='New Improvisation tool + chords_mod_For_YNB.xlsx'
 

scales_df=None
rythmics_df=None
phrases_df=None

def load_xlsx():
    
    global scales_df
    global rythmics_df
    global phrases_df
    # only read specific columns from an excel file
    require_cols = ['Type','Nazwa skali','Scale_len','Tony']
    scales_df = pd.read_excel(fname, \
                              sheet_name='Skale', usecols = require_cols) 
    scales_df=scales_df[['Type','Nazwa skali','Scale_len','Tony']]
#     display(scales_df)
    # rythmics_df = pd.read_excel(fname, \
    #                           sheet_name='Rytmika',header= 1, 
    #                             usecols = [cols for cols in range(69)]).fillna('')
    phrases_df = pd.read_excel(fname, sheet_name='Ogrywanie skal',header= 1, usecols = [cols for cols in range(69)]).fillna('')

def phrases():
    def pcs_phrases():
        global phrases_dfx
        start_col_idx = phrases_df.columns.get_loc('Step vs P1')
        phrases_df.insert(loc = start_col_idx+1,column = 'Pattern',value=None)
        pattern_col_idx = phrases_df.columns.get_loc('Pattern')
#         display(phrases_df)
        # phrases_dfx=phrases_df.head(5).copy()
        phrases_dfx=phrases_df.copy()


        # display(phrases_dfx)
        # for x,y in xsub_df.iterrows():
        #   print(x,y)

#         print(start_col_idx)
        for i in range(phrases_dfx.shape[0]):
            clearFlag=False
            for j in range(start_col_idx,phrases_dfx.shape[1]):
                if phrases_dfx.iloc[i, j]=='x' or clearFlag:
                    phrases_dfx.iloc[i, j]=None
                    clearFlag=True

            pattern=[interval for interval in phrases_dfx.iloc[i, start_col_idx+1:] if interval!=None]
#             print(pattern)
            phrases_dfx.iat[i,pattern_col_idx]=pattern

    def write_wrk_phrases():
        # Getting the current date and time
        dt = datetime.now().strftime("_%Y%m%d_%H%M%S")
        filename="wrk_phrases_dict.xlsx"
        try:
            with pd.ExcelWriter(filename) as writer:
                phrases_dfx.to_excel(writer, sheet_name="Phrases", index=False)
        except:
            filename=filename.split('.')[0]+ dt+'.xlsx'
            with pd.ExcelWriter(filename) as writer:
                phrases_dfx.to_excel(writer, sheet_name="Phrases", index=False)                        
            
    pcs_phrases()

    display(phrases_dfx)
#     print(1)
    # pprint.pprint(phrases_dfx.loc[:,:'Pattern'].to_dict(orient="index"))
#     print(2)
    # preferred
#     phrases_dict=phrases_dfx.loc[:,:'Pattern'].to_dict(orient="records")
    # pprint.pprint(phrases_dict)
    write_wrk_phrases()

def scales():
    def pcs_scales():
        global scales_dfx
        scales_dfx=scales_df.copy()
#         print('blah')
        tones_list_col_idx = scales_dfx.columns.get_loc('Tony')
        scales_dfx.insert(loc=tones_list_col_idx,column = 'Tones_list',value=None)

        for i in range(scales_dfx.shape[0]):
            scales_dfx.iat[i,tones_list_col_idx]=[scale_step for scale_step in scales_dfx.iloc[i, tones_list_col_idx+1].split(' ') \
                          if scale_step!='']

        display(scales_dfx)
    pcs_scales()


In [2]:

    
# https://www.twilio.com/blog/working-with-midi-data-in-python-using-mido    
# https://mido.readthedocs.io/en/latest/
# https://pypi.org/project/isobar/
# https://github.com/ideoforms/isobar/tree/master/examples
# https://www.dummies.com/article/technology/software/music-recording-software/general-music-recording-software/midi-time-code-and-midi-clock-explained-179973/

def main():


    load_xlsx()
    phrases()
    scales()


        
if __name__== "__main__":
    main()
    print("Processing Done")

Def       Type                        Pattern name Step vs P1  \
0    Def Man    General                      Manual pattern         -1   
1    Def 8th  Ogrywanie                       Proste w gore          8   
2    Def 8th  Ogrywanie                        Proste w dol         -8   
3    Def 8th  Ogrywanie                   sec z dolu w gore          4   
4    Def 8th  Ogrywanie                    sec z dolu w dol         -4   
..       ...        ...                                 ...        ...   
422     B5+1    Ogr+bas     b+(5 v 5/4 kwin scale down down    [-1,-1]   
423     B5+1    Ogr+bas    b+(5 v 5/4 terc up-down scale up      [1,1]   
424     B5+1    Ogr+bas  b+(5 v 5/4 terc up-down scale down    [-1,-1]   
425     B5+1    Ogr+bas    b+(5 v 5/4 terc down-up scale up      [1,1]   
426     B5+1    Ogr+bas  b+(5 v 5/4 terc down-up scale down    [-1,-1]   

                             Pattern  P1  P2  P3  P4  P5  ...   P56   P57  \
0        [0, -1, 2, 2, -1, 1, 2, -1]   0  -1   2   2  -1  ...  None  None   
1           [0, 1, 2, 3, 4, 5, 6, 7]   0   1   2   3   4  ...  None  None   
2    [0, -1, -2, -3, -4, -5, -6, -7]   0  -1  -2  -3  -4  ...  None  None   
3           [0, 1, 1, 2, 2, 3, 3, 4]   0   1   1   2   2  ...  None  None   
4      [0, 1, -1, 0, -2, -1, -3, -2]   0   1  -1   0  -2  ...  None  None   
..                               ...  ..  ..  ..  ..  ..  ...   ...   ...   
422  [-5, 0, , -1, , -2, , -3, , -4]  -5   0      -1      ...  None  None   
423      [-5, 0, , 1, , 2, , 1, , 0]  -5   0       1      ...  None  None   
424      [-5, 0, , 1, , 2, , 1, , 0]  -5   0       1      ...  None  None   
425   [-5, 0, , -1, , -2, , -1, , 0]  -5   0      -1      ...  None  None   
426   [-5, 0, , -1, , -2, , -1, , 0]  -5   0      -1      ...  None  None   

      P58   P59   P60   P61   P62   P63   P64 Unnamed: 68  
0    None  None  None  None  None  None  None        None  
1    None  None  None  None  None  None  None        None  
2    None  None  None  None  None  None  None        None  
3    None  None  None  None  None  None  None        None  
4    None  None  None  None  None  None  None        None  
..    ...   ...   ...   ...   ...   ...   ...         ...  
422  None  None  None  None  None  None  None        None  
423  None  None  None  None  None  None  None        None  
424  None  None  None  None  None  None  None        None  
425  None  None  None  None  None  None  None        None  
426  None  None  None  None  None  None  None        None  

[427 rows x 70 columns]

Type                                        Nazwa skali  Scale_len  \
0   Scale                                Dur (Jońska) - 1mod          7   
1   Scale                           durowa miekka (hinduska)          7   
2   Scale                              Mol (Dorycka) - 2 mod          7   
3   Scale                               moll melodyczna down          7   
4   Scale                                 moll melodyczna up          7   
..    ...                                                ...        ...   
75  Chord                  6 molowy Xm11 z sotymą i undecymą          6   
76  Chord  4 zawieszony X7sus4 (septymowy z kwartką zamia...          4   
77  Chord  5 zawieszony X9sus4 (nonowy z kwartką zamiast ...          5   
78  Chord                               4 zmniejszony X7dim4          4   
79  Chord                            4 półzmiejszony X7dim 2          4   

                     Tones_list                      Tony  
0   [1, 3, 5, 6, 8, 10, 12, 13]        1 3 5 6 8 10 12 13  
1    [1, 3, 5, 6, 8, 9, 11, 13]         1 3 5 6 8 9 11 13  
2   [1, 3, 4, 6, 8, 10, 11, 13]        1 3 4 6 8 10 11 13  
3    [1, 3, 4, 6, 8, 9, 11, 13]   1 3 4 6 8 9 11 13        
4   [1, 3, 4, 6, 8, 10, 12, 13]  1 3 4 6 8 10 12 13        
..                          ...                       ...  
75        [1, 4, 8, 11, 15, 18]       1 4 8 11 15 18       
76                [1, 6, 8, 11]         1 6 8 11           
77            [1, 6, 8, 11, 15]        1 6 8 11 15         
78                [1, 4, 7, 10]         1 4 7 10           
79                [1, 4, 7, 11]         1 4 7 11           

[80 rows x 5 columns]

Processing Done


MidiFile(type=0, ticks_per_beat=480  (or ticks per quarater note or PPQ parts per quarter), tracks=[
  MidiTrack([
    MetaMessage('text', text='note track', time=0),
    MetaMessage('set_tempo', tempo=500000, time=0), tempo=expressed in microseconds (1/milionth) this means 0.5 sec., which gives 120 bpm
    MetaMessage('key_signature', key='C', time=0),
    MetaMessage('time_signature', numerator=4, denominator=4, clocks_per_click=48, notated_32nd_notes_per_beat=8, time=0),

notated_32nd_notes_per_beat=8 -> this means per quarter, and actually this clarifies beat as quarter here (could be different thought, where beat is not quarter)

ticks_per_beat=480  (resolution)
beats_per_minute=120 (tempo)   (expressed as 0,5s = 500.000 (tempo))

ticks per minute would be 57600

Midi time is expressed in ticks.!!!

What the PPQ means in terms of absolute time depends on the designated tempo.


The file header specifies the number of ticks per beat, and the tempo messages specify the length of a beat, in microseconds. You can compute a delta time value in microseconds with tempo * delta_ticks / ticks_per_beat, but you have to take care to merge all tracks first because a tempo message in another track can change that value at any time

tempo here is 4/4 3/4

3/4
MidiFile(type=0, ticks_per_beat=480, tracks=[
  MidiTrack([
    MetaMessage('text', text='note track', time=0),
    MetaMessage('set_tempo', tempo=500000, time=0),
    MetaMessage('key_signature', key='C', time=0),
    MetaMessage('time_signature', numerator=3, denominator=4, clocks_per_click=18, notated_32nd_notes_per_beat=8, time=0),
    
MidiFile(type=0, ticks_per_beat=480, tracks=[
  MidiTrack([
    MetaMessage('text', text='note track', time=0),
    MetaMessage('set_tempo', tempo=500000, time=0),
    MetaMessage('key_signature', key='C', time=0),
    MetaMessage('time_signature', numerator=3, denominator=4, clocks_per_click=18, notated_32nd_notes_per_beat=8, time=0),    
    
    
    

In [104]:
del mid

In [18]:
names=mido.get_output_names()
print(names)
port = mido.open_output('Microsoft GS Wavetable Synth 0')
# port.close()
print(port)

['Microsoft GS Wavetable Synth 0', 'Bome Virtual MIDI Port 1', 'Bome Virtual MIDI Port 2']
<open output 'Microsoft GS Wavetable Synth 0' (RtMidi/WINDOWS_MM)>


In [41]:
print(port)


<open output 'Microsoft GS Wavetable Synth 0' (RtMidi/WINDOWS_MM)>
{'name': 'Microsoft GS Wavetable Synth 0', '_lock': <mido.ports.DummyLock object at 0x0000016F5838E100>, 'closed': False, '_send_lock': <unlocked _thread.RLock object owner=0 count=0 at 0x0000016F58236A80>, '_rt': <rtmidi._rtmidi.MidiOut object at 0x0000016F5F962040>, 'api': 'WINDOWS_MM', '_device_type': 'RtMidi/WINDOWS_MM', 'autoreset': False}


## Direct write of events into midi output port

In [5]:
import mido
import time

def direct_midi_play():
    names=mido.get_output_names()
    print(names)

    port = mido.open_output('Microsoft GS Wavetable Synth 0') 
    # port.close()

    print(port)
    # mid=mido.MidiFile('Def 5 4.mid',clip=True)
    # mid=mido.MidiFile('rythm_midi_files\\Def (5.mid',clip=True)
    mid=mido.MidiFile('rythm_midi_files\\Dim6_2.mid',clip=True)
    # mid=mido.MidiFile('rythm_midi_files\\Blah.mid',clip=True)
    # mid=mido.MidiFile('rythm_midi_files\\Blah2.mid',clip=True)
    for msg in mid:
        time.sleep(msg.time)
        if not msg.is_meta:
            port.send(msg)

    port.close()   

direct_midi_play()

['Microsoft GS Wavetable Synth 0', 'Bome Virtual MIDI Port 1', 'Bome Virtual MIDI Port 2']
<open output 'Microsoft GS Wavetable Synth 0' (RtMidi/WINDOWS_MM)>


In [40]:
for msg in mid.tracks[0]:
    if msg.type=='time_signature':
        print(msg)
        msg.clocks_per_click=12
for msg in mid:
    time.sleep(msg.time)
#     port.send(msg)
    if not msg.is_meta:
        port.send(msg)

for msg in mid.tracks[0]:
    if msg.type=='time_signature':
        print(msg)
        msg.clocks_per_click=48
for msg in mid:
    time.sleep(msg.time)
#     port.send(msg)
    if not msg.is_meta:
        port.send(msg)
        
        

MetaMessage('time_signature', numerator=3, denominator=4, clocks_per_click=48, notated_32nd_notes_per_beat=8, time=0)


ValueError: 'message' longer than 3 bytes but does not start with 0xF0.

## dump content of midi file

In [8]:
def dump_midi_file():
    mid=mido.MidiFile('rythm_midi_files\\Dim6_2.mid',clip=True)
    print(mid)
    print('xxx')
    for track in mid.tracks:
        print(track)
    print('yyy')
    # FIXME: blah

    for msg in mid.tracks[0]:
        print(msg)
        # TODO: blah blah
    
dump_midi_file()    
   
    
    

MidiFile(type=0, ticks_per_beat=480, tracks=[
  MidiTrack([
    MetaMessage('text', text='note track', time=0),
    MetaMessage('set_tempo', tempo=500000, time=0),
    MetaMessage('key_signature', key='C', time=0),
    MetaMessage('time_signature', numerator=4, denominator=4, clocks_per_click=48, notated_32nd_notes_per_beat=8, time=0),
    Message('control_change', channel=0, control=7, value=96, time=0),
    Message('control_change', channel=0, control=10, value=64, time=0),
    Message('program_change', channel=0, program=0, time=1),
    MetaMessage('track_name', name='Dim6_2', time=0),
    Message('control_change', channel=0, control=7, value=96, time=0),
    Message('control_change', channel=0, control=110, value=0, time=0),
    Message('control_change', channel=0, control=111, value=0, time=0),
    Message('control_change', channel=0, control=112, value=13, time=0),
    Message('control_change', channel=0, control=113, value=0, time=0),
    Message('control_change', channel=0, con

## isobar play (background) example

In [7]:
# isobar play (background) example
import isobar as iso
import os
from isobar.io import MidiFileInputDevice

import logging
logging.basicConfig(level=logging.INFO, format="[%(asctime)s] %(message)s")

def play_test():
    print(os.getenv("ISOBAR_DEFAULT_MIDI_OUT"))
    name = "Microsoft GS Wavetable Synth 0"
    if "midi_out" not in locals():
        midi_out = iso.MidiOutputDevice(device_name=name,
                                        send_clock=True)

    # pattern = MidiFileInputDevice('rythm_midi_files\\Blah2.mid').read()
    pattern = MidiFileInputDevice('rythm_midi_files\\Dim6_2.mid').read()
    print("Read pattern containing %d note events" % len(pattern["note"]))
    print(pattern)

    # mid=mido.MidiFile('rythm_midi_files\\Dim6_2x.mid',clip=True)
    # mid=mido.MidiFile('rythm_midi_files\\Blah.mid',clip=True)
    # mid=mido.MidiFile('rythm_midi_files\\Blah2.mid',clip=True)
    # del timeline
    timeline = iso.Timeline(120,output_device=midi_out)
#     timeline = iso.Timeline(120)
    stream=timeline.schedule(pattern)
    # timeline.stop_when_done=False
    # timeline.stop_when_done=True
    print(dir(timeline))
    print(timeline.tick_duration, timeline.ticks_per_beat, timeline.beats_to_seconds(123), timeline.seconds_to_beats(1.5))
    timeline.set_tempo=100000
    timeline.background()  # use background instead of run to enable live performing (async notes passing)
#     timeline.unschedule(stream)
    return timeline

def play_cont_test(internal_timeline):
    pattern = MidiFileInputDevice('rythm_midi_files\\Dim6_1.mid').read()
    pattern = MidiFileInputDevice('rythm_midi_files\\Dim3.mid').read()
    # pattern = MidiFileInputDevice('rythm_midi_files\\Dim4_1.mid').read()

    # stream.update({
    #     "note" : iso.PSequence([ 48 ]),
    #     "duration": iso.PSequence([ 1 ])
    # }, quantize=1)
    # stream.update(pattern)
    stream=internal_timeline.schedule(pattern)

tm_tst=play_test()    

[2022-05-21 22:09:56,629] Clock: Timer overflowed (late by 0.027s)
[2022-05-21 22:09:56,759] Clock: Timer overflowed (late by 0.128s)


None


DeviceNotFoundException: Could not find MIDI device

[2022-05-21 22:09:56,824] Clock: Timer overflowed (late by 0.063s)
[2022-05-21 22:10:20,769] Clock: Timer overflowed (late by 0.025s)
[2022-05-21 22:10:20,791] Clock: Timer overflowed (late by 0.021s)
[2022-05-21 22:10:22,901] Clock: Timer overflowed (late by 0.021s)
[2022-05-21 22:10:22,926] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:10:31,383] Clock: Timer overflowed (late by 0.023s)
[2022-05-21 22:10:31,429] Clock: Timer overflowed (late by 0.021s)
[2022-05-21 22:10:35,066] Clock: Timer overflowed (late by 0.027s)
[2022-05-21 22:10:35,173] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:10:35,199] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:10:37,863] Clock: Timer overflowed (late by 0.029s)
[2022-05-21 22:10:38,471] Clock: Timer overflowed (late by 0.021s)
[2022-05-21 22:10:38,498] Clock: Timer overflowed (late by 0.025s)
[2022-05-21 22:10:38,579] Clock: Timer overflowed (late by 0.080s)
[2022-05-21 22:10:38,621] Clock: Timer overflowed (late by 0.0

In [26]:
del timeline
del midi_out

[2022-05-14 22:22:24,374] Clock: Timer overflowed (late by 0.033s)
[2022-05-14 22:22:24,393] Clock: Timer overflowed (late by 0.037s)


In [14]:
# wrk
import isobar as iso
import os
from isobar.io import MidiFileInputDevice

import logging
logging.basicConfig(level=logging.INFO, format="[%(asctime)s] %(message)s")

print(os.getenv("ISOBAR_DEFAULT_MIDI_OUT"))
name = "Microsoft GS Wavetable Synth 0"
if "midi_out" not in locals():
    midi_out = iso.MidiOutputDevice(device_name=name,
                                    send_clock=True)

# pattern = MidiFileInputDevice('rythm_midi_files\\Blah2.mid').read()
pattern = MidiFileInputDevice('rythm_midi_files\\Dim6_2.mid').read()
print("Read pattern containing %d note events" % len(pattern["note"]))
print(pattern)

# mid=mido.MidiFile('rythm_midi_files\\Dim6_2x.mid',clip=True)
# mid=mido.MidiFile('rythm_midi_files\\Blah.mid',clip=True)
# mid=mido.MidiFile('rythm_midi_files\\Blah2.mid',clip=True)
# del timeline
timeline = iso.Timeline(120,output_device=midi_out)
stream=timeline.schedule(pattern)
# timeline.stop_when_done=False
# timeline.stop_when_done=True
print(dir(timeline))
print(timeline.tick_duration, timeline.ticks_per_beat, timeline.beats_to_seconds(123), timeline.seconds_to_beats(1.5))
timeline.set_tempo=100000
timeline.background()  # use background instead of run to enable live performing (async notes passing)
# pprint.pprint(pattern)
# for note in pattern['note']:
#     print(note)
# print('x')    
# for note in pattern['amplitude']:
#     print(note)
# print('x')    
# for note in pattern['gate']:
#     print(note)
# print('x')    
# for note in pattern['duration']:
#     print(note)
    


    

[2022-05-14 22:25:31,855] Loading MIDI data from rythm_midi_files\Dim6_2.mid, ticks per beat = 480
[2022-05-14 22:25:31,857] Timeline: Scheduled new track (total tracks: 1)


None
Read pattern containing 6 note events
{'note': <isobar.pattern.sequence.PSequence object at 0x0000025B526A4E20>, 'amplitude': <isobar.pattern.sequence.PSequence object at 0x0000025B528D7070>, 'gate': <isobar.pattern.sequence.PSequence object at 0x0000025B528D7100>, 'duration': <isobar.pattern.sequence.PSequence object at 0x0000025B528D7190>}
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_clock_source', 'add_output_device', 'background', 'beats_to_seconds', 'clock_multipliers', 'clock_source', 'current_time', 'dump', 'events', 'get_clock_source', 'get_output_device', 'get_tempo', 'get_ticks_per_beat', 'max_tracks', 'output_device', 'output_devices', 'reset', 'reset_to_beat', 'run', 'runn

[2022-05-14 22:25:31,864] Timeline: Starting


In [31]:
pattern = MidiFileInputDevice('rythm_midi_files\\Dim6_1.mid').read()
pattern = MidiFileInputDevice('rythm_midi_files\\Dim3.mid').read()
# pattern = MidiFileInputDevice('rythm_midi_files\\Dim4_1.mid').read()

# stream.update({
#     "note" : iso.PSequence([ 48 ]),
#     "duration": iso.PSequence([ 1 ])
# }, quantize=1)
# stream.update(pattern)
stream=timeline.schedule(pattern) #timeline schedule or stream update to give new elemennts to play
#It seems that update could be better for memory.

# timeline.unschedule()
# timeline.stop()

[2022-05-14 22:27:46,432] Loading MIDI data from rythm_midi_files\Dim6_1.mid, ticks per beat = 480
[2022-05-14 22:27:46,438] Loading MIDI data from rythm_midi_files\Dim3.mid, ticks per beat = 480
[2022-05-14 22:27:46,440] Timeline: Scheduled new track (total tracks: 2)
[2022-05-14 22:27:49,797] Timeline: Track finished, removing from scheduler (total tracks: 1)
[2022-05-14 22:27:55,885] Clock: Timer overflowed (late by 0.021s)
[2022-05-14 22:27:55,942] Clock: Timer overflowed (late by 0.078s)
[2022-05-14 22:27:55,943] Clock: Timer overflowed (late by 0.079s)
[2022-05-14 22:27:55,944] Clock: Timer overflowed (late by 0.057s)
[2022-05-14 22:27:56,007] Timeline: Track finished, removing from scheduler (total tracks: 0)
[2022-05-14 22:27:56,022] Clock: Timer overflowed (late by 0.046s)
[2022-05-14 22:27:56,022] Clock: Timer overflowed (late by 0.046s)
[2022-05-14 22:27:56,022] Clock: Timer overflowed (late by 0.077s)


In [32]:
# timeline.unschedule()
timeline.stop()


[2022-05-14 22:27:56,851] Timeline: Stopping
[2022-05-14 22:28:00,078] Clock: Timer overflowed (late by 0.029s)
[2022-05-14 22:28:02,895] Clock: Timer overflowed (late by 0.020s)
[2022-05-14 22:28:02,921] Clock: Timer overflowed (late by 0.025s)
[2022-05-14 22:28:08,129] Clock: Timer overflowed (late by 0.022s)
[2022-05-14 22:28:08,152] Clock: Timer overflowed (late by 0.021s)
[2022-05-14 22:28:13,893] Clock: Timer overflowed (late by 0.020s)
[2022-05-14 22:28:17,258] Clock: Timer overflowed (late by 0.020s)
[2022-05-14 22:28:41,469] Clock: Timer overflowed (late by 0.020s)
[2022-05-14 22:28:41,492] Clock: Timer overflowed (late by 0.021s)
[2022-05-14 22:28:41,532] Clock: Timer overflowed (late by 0.023s)
[2022-05-14 22:28:41,538] Clock: Timer overflowed (late by 0.029s)
[2022-05-14 22:28:41,555] Clock: Timer overflowed (late by 0.021s)
[2022-05-14 22:28:52,261] Clock: Timer overflowed (late by 0.020s)
[2022-05-14 22:28:52,302] Clock: Timer overflowed (late by 0.026s)
[2022-05-14 22:28

[2022-05-14 22:30:20,470] Clock: Timer overflowed (late by 0.038s)
[2022-05-14 22:30:20,476] Clock: Timer overflowed (late by 0.033s)
[2022-05-14 22:30:20,535] Clock: Timer overflowed (late by 0.022s)
[2022-05-14 22:30:20,559] Clock: Timer overflowed (late by 0.022s)
[2022-05-14 22:30:20,882] Clock: Timer overflowed (late by 0.021s)
[2022-05-14 22:30:21,075] Clock: Timer overflowed (late by 0.023s)
[2022-05-14 22:30:21,077] Clock: Timer overflowed (late by 0.025s)
[2022-05-14 22:30:21,098] Clock: Timer overflowed (late by 0.022s)
[2022-05-14 22:30:21,396] Clock: Timer overflowed (late by 0.042s)
[2022-05-14 22:30:21,481] Clock: Timer overflowed (late by 0.083s)
[2022-05-14 22:30:21,529] Clock: Timer overflowed (late by 0.047s)
[2022-05-14 22:30:21,545] Clock: Timer overflowed (late by 0.125s)
[2022-05-14 22:30:21,569] Clock: Timer overflowed (late by 0.023s)
[2022-05-14 22:30:21,591] Clock: Timer overflowed (late by 0.045s)
[2022-05-14 22:30:21,622] Clock: Timer overflowed (late by 0.0

[2022-05-14 22:30:38,618] Clock: Timer overflowed (late by 0.030s)
[2022-05-14 22:30:38,703] Clock: Timer overflowed (late by 0.020s)
[2022-05-14 22:30:38,793] Clock: Timer overflowed (late by 0.031s)
[2022-05-14 22:30:38,794] Clock: Timer overflowed (late by 0.031s)
[2022-05-14 22:30:38,825] Clock: Timer overflowed (late by 0.029s)
[2022-05-14 22:30:38,835] Clock: Timer overflowed (late by 0.041s)
[2022-05-14 22:30:38,857] Clock: Timer overflowed (late by 0.030s)
[2022-05-14 22:30:38,880] Clock: Timer overflowed (late by 0.044s)
[2022-05-14 22:30:39,404] Clock: Timer overflowed (late by 0.033s)
[2022-05-14 22:30:39,432] Clock: Timer overflowed (late by 0.026s)
[2022-05-14 22:31:10,192] Clock: Timer overflowed (late by 0.021s)
[2022-05-14 22:31:10,211] Clock: Timer overflowed (late by 0.024s)
[2022-05-14 22:31:10,219] Clock: Timer overflowed (late by 0.026s)
[2022-05-14 22:31:49,489] Clock: Timer overflowed (late by 0.023s)
[2022-05-14 22:31:49,508] Clock: Timer overflowed (late by 0.0

In [5]:
play_cont_test(tm_tst)

[2022-05-14 22:44:12,593] Loading MIDI data from rythm_midi_files\Dim6_1.mid, ticks per beat = 480
[2022-05-14 22:44:12,598] Loading MIDI data from rythm_midi_files\Dim3.mid, ticks per beat = 480
[2022-05-14 22:44:12,599] Timeline: Scheduled new track (total tracks: 1)
[2022-05-14 22:44:22,099] Timeline: Track finished, removing from scheduler (total tracks: 0)


In [6]:
# tm_tst.unschedule()
tm_tst.stop()

[2022-05-14 22:44:28,856] Timeline: Stopping


In [18]:
import isobar as iso
import time

def print_message(message):
    """
    The callback argument is a mido Message object:
    https://mido.readthedocs.io/en/latest/messages.html
    """
    print(" - Received MIDI: %s" % message)

midi_in = iso.MidiInputDevice()
midi_in.callback = print_message
print("Opened MIDI input: %s" % midi_in.device_name)

try:
    while True:
        time.sleep(0.1)
except KeyboardInterrupt:
    pass


[2022-05-14 21:53:15,115] Opened MIDI input: Bome Virtual MIDI Port 0


Opened MIDI input: Bome Virtual MIDI Port 0


## load all rythm pattern midis wth mido and isobar

In [9]:
# load all rythm pattern midis with mido and isobar
import isobar as iso
import os
from isobar.io import MidiFileInputDevice
import time
import mido

import logging
logging.basicConfig(level=logging.INFO, format="[%(asctime)s] %(message)s")

midi_files = ['B2+1.mid',
'B3+1.mid',
'B4+1.mid',
'B5+1_(5.mid',
'B5+1_5 4.mid',
'Chord3.mid',
'Chord4.mid',
'Chord5.mid',
'Chord6.mid',
'Chord7.mid',
'Dec (3.mid',
'Def (5.mid',
'Def 5 4.mid',
'Def8.mid',
'Dim2.mid',
'Dim3.mid',
'Dim4_1.mid',
'Dim4_2.mid',
'Dim5.mid',
'Dim6_1.mid',
'Dim6_2.mid',
'Dim7.mid',
'Dim8.mid',
'Dim9.mid',
'Orn grace.mid',
'Orn kwin.mid',
'Orn mord.mid',
'Orn quat.mid',
'Orn trio.mid']

all_midis=[]
for midi_file in midi_files:
    print(midi_file)
    all_midis.append({'name':midi_file, \
                      'isobar_midi':MidiFileInputDevice('rythm_midi_files\\'+midi_file).read(),\
                      'mid':mido.MidiFile('rythm_midi_files\\'+midi_file)
                     }
                    )

print(all_midis)    

[2022-05-15 20:12:41,419] Loading MIDI data from rythm_midi_files\B2+1.mid, ticks per beat = 480
[2022-05-15 20:12:41,438] Loading MIDI data from rythm_midi_files\B3+1.mid, ticks per beat = 480
[2022-05-15 20:12:41,468] Loading MIDI data from rythm_midi_files\B4+1.mid, ticks per beat = 480
[2022-05-15 20:12:41,505] Loading MIDI data from rythm_midi_files\B5+1_(5.mid, ticks per beat = 480
[2022-05-15 20:12:41,526] Loading MIDI data from rythm_midi_files\B5+1_5 4.mid, ticks per beat = 480
[2022-05-15 20:12:41,548] Loading MIDI data from rythm_midi_files\Chord3.mid, ticks per beat = 480
[2022-05-15 20:12:41,564] Loading MIDI data from rythm_midi_files\Chord4.mid, ticks per beat = 480
[2022-05-15 20:12:41,571] Loading MIDI data from rythm_midi_files\Chord5.mid, ticks per beat = 480
[2022-05-15 20:12:41,579] Loading MIDI data from rythm_midi_files\Chord6.mid, ticks per beat = 480
[2022-05-15 20:12:41,592] Loading MIDI data from rythm_midi_files\Chord7.mid, ticks per beat = 480
[2022-05-15 2

B2+1.mid
B3+1.mid
B4+1.mid
B5+1_(5.mid
B5+1_5 4.mid
Chord3.mid
Chord4.mid
Chord5.mid
Chord6.mid
Chord7.mid
Dec (3.mid


[2022-05-15 20:12:41,636] Loading MIDI data from rythm_midi_files\Def (5.mid, ticks per beat = 480
[2022-05-15 20:12:41,651] Loading MIDI data from rythm_midi_files\Def 5 4.mid, ticks per beat = 480
[2022-05-15 20:12:41,677] Loading MIDI data from rythm_midi_files\Def8.mid, ticks per beat = 480
[2022-05-15 20:12:41,705] Loading MIDI data from rythm_midi_files\Dim2.mid, ticks per beat = 480
[2022-05-15 20:12:41,713] Loading MIDI data from rythm_midi_files\Dim3.mid, ticks per beat = 480
[2022-05-15 20:12:41,728] Loading MIDI data from rythm_midi_files\Dim4_1.mid, ticks per beat = 480
[2022-05-15 20:12:41,740] Loading MIDI data from rythm_midi_files\Dim4_2.mid, ticks per beat = 480
[2022-05-15 20:12:41,753] Loading MIDI data from rythm_midi_files\Dim5.mid, ticks per beat = 480
[2022-05-15 20:12:41,763] Loading MIDI data from rythm_midi_files\Dim6_1.mid, ticks per beat = 480
[2022-05-15 20:12:41,772] Loading MIDI data from rythm_midi_files\Dim6_2.mid, ticks per beat = 480
[2022-05-15 20:12

Def (5.mid
Def 5 4.mid
Def8.mid
Dim2.mid
Dim3.mid
Dim4_1.mid
Dim4_2.mid
Dim5.mid
Dim6_1.mid
Dim6_2.mid
Dim7.mid
Dim8.mid
Dim9.mid
Orn grace.mid


[2022-05-15 20:12:41,828] Loading MIDI data from rythm_midi_files\Orn kwin.mid, ticks per beat = 480
[2022-05-15 20:12:41,837] Loading MIDI data from rythm_midi_files\Orn mord.mid, ticks per beat = 480
[2022-05-15 20:12:41,849] Loading MIDI data from rythm_midi_files\Orn quat.mid, ticks per beat = 480
[2022-05-15 20:12:41,861] Loading MIDI data from rythm_midi_files\Orn trio.mid, ticks per beat = 480


Orn kwin.mid
Orn mord.mid
Orn quat.mid
Orn trio.mid
[{'name': 'B2+1.mid', 'isobar_midi': {'note': <isobar.pattern.sequence.PSequence object at 0x000001D7EE7BC550>, 'amplitude': <isobar.pattern.sequence.PSequence object at 0x000001D7EF2BCDC0>, 'gate': <isobar.pattern.sequence.PSequence object at 0x000001D7EF2BCE20>, 'duration': <isobar.pattern.sequence.PSequence object at 0x000001D7EF2BCE80>}, 'mid': MidiFile(type=0, ticks_per_beat=480, tracks=[
  MidiTrack([
    MetaMessage('text', text='note track', time=0),
    MetaMessage('set_tempo', tempo=500000, time=0),
    MetaMessage('key_signature', key='C', time=0),
    MetaMessage('time_signature', numerator=4, denominator=4, clocks_per_click=48, notated_32nd_notes_per_beat=8, time=0),
    Message('control_change', channel=0, control=7, value=96, time=0),
    Message('control_change', channel=0, control=10, value=64, time=0),
    Message('program_change', channel=0, program=0, time=1),
    MetaMessage('track_name', name='B2+1', time=0),
   

In [100]:
## Check midis content
for midi_data in all_midis:
#     if midi_data['name']!='Def8.mid':
    if midi_data['name']!='Dec (3.mid':
        continue
    print(midi_data['name'])

    pattern=midi_data['isobar_midi']
#     print(pattern)
#     print(pattern['note'].__dict__)
#     print(dir(pattern['note']))
#     print(pattern.all())
#     print(pattern['amplitude'].__dict__)
    print(pattern['duration'].sequence, sum(pattern['duration'].sequence), len(pattern['duration'].sequence))


    for note,amplitude,gate,duration in zip(pattern['note'].sequence,pattern['amplitude'].sequence,\
                                            pattern['gate'].sequence,pattern['duration'].sequence):
        pass
        print({'note':note,'amplitude':amplitude,'gate':gate,'duration':duration })
#     break
    xxx=zip(pattern['note'].sequence,pattern['amplitude'].sequence,pattern['gate'].sequence,pattern['duration'].sequence)
#     print(dir(xxx))
    break
    
    
def dump_midi_file():
#     mid=mido.MidiFile('rythm_midi_files\\Dim6_2.mid',clip=True)
#     mid=mido.MidiFile('rythm_midi_files\\Orn trio.mid',clip=True)
    mid=mido.MidiFile('rythm_midi_files\\Def8.mid',clip=True)
#     print(mid)
    print('xxx')
    print("ticks_per_beat",mid.ticks_per_beat)
    for msg in mid.tracks[0]:
        if msg.is_meta:
            print('a:',msg)
        elif 'note' in msg.type:
            print('b:',msg.time,msg.time/mid.ticks_per_beat)
            print('c:',1+msg.time,(1+msg.time)/mid.ticks_per_beat)
            print('aaa:',msg.__dict__)


    len_in_tick=sum([msg.time for msg in mid.tracks[0]])
    print(len_in_tick)
    print(len_in_tick/mid.ticks_per_beat)
    print((len_in_tick-26)/mid.ticks_per_beat)
#     print(msg.time)
    
    
dump_midi_file()    
       

Dec (3.mid
[0.6666666666666666, 0.6666666666666666, 0.6666666666666667, 0.6666666666666665, 0.6666666666666665, 0.6666666666666665, 0.6666666666666665, 0.6666666666666661, 0.6666666666666661, 0.6666666666666661, 0.6666666666666661, 0.6645833333333329] 7.997916666666663 12
{'note': 61, 'amplitude': 105, 'gate': 0.996875, 'duration': 0.6666666666666666}
{'note': 64, 'amplitude': 80, 'gate': 0.9968749999999998, 'duration': 0.6666666666666666}
{'note': 68, 'amplitude': 80, 'gate': 0.9968749999999998, 'duration': 0.6666666666666667}
{'note': 64, 'amplitude': 95, 'gate': 0.9968750000000002, 'duration': 0.6666666666666665}
{'note': 68, 'amplitude': 80, 'gate': 0.9968750000000002, 'duration': 0.6666666666666665}
{'note': 73, 'amplitude': 80, 'gate': 0.9968750000000002, 'duration': 0.6666666666666665}
{'note': 68, 'amplitude': 95, 'gate': 0.9968750000000002, 'duration': 0.6666666666666665}
{'note': 73, 'amplitude': 80, 'gate': 0.9968750000000002, 'duration': 0.6666666666666661}
{'note': 76, 'am

In [37]:
import pprint
for midi_data in all_midis:
#     print(midi_data['mid'])
    mid=midi_data['mid']


#     for track in mid.tracks:
#         print(track)
    
    for msg in mid.tracks[0]:
#         print(msg.__dict__)
#         {'type': 'control_change', 'time': 0, 'control': 10, 'value': 64, 'channel': 0}
#         pprint.pprint(dir(msg))

        if msg.type=='control_change' and msg.value!=0:
#             print('asdf ',msg.value)
#             msg.remove()
            print(msg)
        # TODO: blah blah


control_change channel=0 control=7 value=96 time=0
control_change channel=0 control=10 value=64 time=0
control_change channel=0 control=7 value=96 time=0
control_change channel=0 control=112 value=13 time=0
control_change channel=0 control=114 value=2 time=0
control_change channel=0 control=112 value=13 time=0
control_change channel=0 control=114 value=11 time=0
control_change channel=0 control=112 value=13 time=0
control_change channel=0 control=114 value=18 time=0
control_change channel=0 control=112 value=13 time=0
control_change channel=0 control=114 value=25 time=0
control_change channel=0 control=112 value=13 time=0
control_change channel=0 control=114 value=32 time=0
control_change channel=0 control=112 value=13 time=0
control_change channel=0 control=114 value=40 time=0
control_change channel=0 control=112 value=13 time=0
control_change channel=0 control=114 value=47 time=0
control_change channel=0 control=112 value=13 time=0
control_change channel=0 control=114 value=55 time=0

CC to check
7 - Controls the volume of the channel.
10 - Controls the left and right balance, generally for mono patches. A value of 64 equals the center. (mono loaded to stereo)
112
114
113


In [81]:
arpeggio = iso.PSeries(0, 2, 6)
for x in arpeggio:
    print(x)
print('x')    
print(arpeggio.__dict__)
arpeggio = iso.PDegree(arpeggio, iso.Scale.minor) + 72
for x in arpeggio:
    print(x)
# print(arpeggio.__dict__)    
# print(arpeggio.a.__dict__)    
# print(arpeggio.a.degree.__dict__)    
# print(arpeggio.a.scale.__dict__)    
# print(arpeggio.b.__dict__)    

# # print('x')    
# arpeggio = iso.PPingPong(arpeggio)
# print(arpeggio.__dict__)    
# # for x in arpeggio:
# #     print(x)
# # print('x')    
# arpeggio = iso.PLoop(arpeggio)
# print(arpeggio.__dict__)    

0
2
4
6
8
10
x
{'start': 0, 'value': 12, 'step': 2, 'length': 6, 'count': 6}
{'a': <isobar.pattern.tonal.PDegree object at 0x0000016F064A4910>, 'b': <isobar.pattern.core.PConstant object at 0x0000016F064A4D30>}
{'degree': <isobar.pattern.sequence.PSeries object at 0x0000016F0651EFA0>, 'scale': <isobar.scale.Scale object at 0x0000016F0568F970>}
{'start': 0, 'value': 12, 'step': 2, 'length': 6, 'count': 6}
{'semitones': [0, 2, 3, 5, 7, 8, 11], 'weights': [0.14285714285714285, 0.14285714285714285, 0.14285714285714285, 0.14285714285714285, 0.14285714285714285, 0.14285714285714285, 0.14285714285714285], 'name': 'minor', 'octave_size': 12}
{'constant': 72}
{'pattern': <isobar.pattern.core.PAdd object at 0x0000016F064312B0>, 'count': 1, 'values': [72, 75, 79, 83, 86, 89], 'pos': 0, 'dir': 1, 'rpos': 0}
{'pattern': <isobar.pattern.sequence.PPingPong object at 0x0000016F0651EF70>, 'count': 9223372036854775807, 'pos': 0, 'loop_index': 0, 'read_all': False, 'values': []}


In [110]:
# print(iso.Scale.__dict__)
# for scale in iso.Scale.__dict__:
#     print(scale)
seq=iso.pattern.PSequence([45, 35, 25, 40], repeats=4)
print(seq.__dict__)
for x in seq:
    print(x)
    
seq.rcount=2
seq.pos=2
print(seq.__dict__)
for x in seq:
    print(x)
    

{'sequence': [45, 35, 25, 40], 'repeats': 4, 'rcount': 0, 'pos': 0}
45
35
25
40
45
35
25
40
45
35
25
40
45
35
25
40
{'sequence': [45, 35, 25, 40], 'repeats': 4, 'rcount': 2, 'pos': 2}
25
40
45
35
25
40


In [11]:
import isobar as iso 
arpeggio = iso.PSeries(0, 2, 6)/3
print(arpeggio)
print([x for x in arpeggio])
# arpeggio2=arpeggio.sequence()/3
# arpeggio2=PAdd(arpeggio,2)
# print([x for x in arpeggio2])

amp = iso.PSequence([40, 30, 20, 25], repeats=1)
amp2=amp/6
print(1234234, len(amp))
print([x for x in amp2])
amp3=iso.PSequence(list(amp),repeats=2)
print([x for x in amp3])



(Pattern (<class 'isobar.pattern.sequence.PSeries'>)) / (constant)
[0.0, 0.6666666666666666, 1.3333333333333333, 2.0, 2.6666666666666665, 3.3333333333333335]
1234234 4
[6.666666666666667, 5.0, 3.3333333333333335, 4.166666666666667]
[40, 30, 20, 25, 40, 30, 20, 25]


# PSequence as foundation for patterns (atom - amplitude, lentgh,height)
PSeries could be good also.

## How to deal with patterns
http://ideoforms.github.io/isobar/patterns/
including copying and resetting

In [46]:
import isobar as iso 

#example of PSeries with math operation
arpeggio = iso.PSeries(0, 2, 6)/3

#example of PSequence
amp = iso.PSequence([40, 30, 20, 25], repeats=1)

# which means would need to be reinstated each time
print('ad1: ',amp.__dict__)
print('amp: ',[x for x in amp])
print('ad2: ',amp.__dict__)
ampp=amp
print('ad3: ',ampp.__dict__)
ampp.rcount=0
print('ad4: ',ampp.__dict__)
# print('amp_: ',[x for x in ampp])
ampp=list(amp)
print('amp_: ',[x for x in ampp])
# print(list(amp.copy()))
print('ampX: ',[x for x in amp])

# math operation
amp2=amp/6
print('amp2: ',[x for x in amp2])

#increasing repeat
amp3=iso.PSequence(list(amp),repeats=2)
print('amp3: ',[x for x in amp3])



ad1:  {'sequence': [40, 30, 20, 25], 'repeats': 1, 'rcount': 0, 'pos': 0}
amp:  [40, 30, 20, 25]
ad2:  {'sequence': [40, 30, 20, 25], 'repeats': 1, 'rcount': 1, 'pos': 0}
ad3:  {'sequence': [40, 30, 20, 25], 'repeats': 1, 'rcount': 1, 'pos': 0}
ad4:  {'sequence': [40, 30, 20, 25], 'repeats': 1, 'rcount': 0, 'pos': 0}
amp_:  [40, 30, 20, 25]
ampX:  []
amp2:  []
amp3:  [40, 30, 20, 25, 40, 30, 20, 25]


# Example with fixed lengs, amps and rising melody

In [1]:
import isobar as iso 

# create 1/8 x 8 pattern with 96 as amplitude and [1,3,2,4,3,5,4,6]
amplitude = iso.PSequence([96,96,96,96], repeats=2)
length = iso.PSequence([1/2,1/2,1/2,1/2], repeats=2)
notes = iso.PSequence([1,3,2,4,3,5,4,6], repeats=1)







In [2]:
# one time init  (this is the best to used globals)
# One time is sufficient, so basically this may be repeated

import isobar as iso
global midi_out
global timeline

name = "Microsoft GS Wavetable Synth 0"
print('mo in gl:','midi_out' in globals())
if 'midi_out' not in locals():
    print('no midiout')
    midi_out = iso.MidiOutputDevice(device_name=name)

else:
    print('yes: midi out')    
    
if 'timeline' not in locals():
    timeline = iso.Timeline(120,output_device=midi_out)    
    

mo in gl: False
no midiout


[2022-05-21 22:56:16,402] Opened MIDI output: Microsoft GS Wavetable Synth 0


In [3]:
# isobar play (background) example
import isobar as iso



import time

import logging
global midi_out

logging.basicConfig(level=logging.INFO, format="[%(asctime)s] %(message)s")



def play_test():
    midi_out
#     timeline = iso.Timeline(120,output_device=midi_out)
    
    
    #--------------------------------------------------------------------------------
    # Play a series of 5 notes with random velocities.
    # Delay by 1 beat before playback.
    #--------------------------------------------------------------------------------
#     timeline.schedule({
#         "note": iso.PSequence([ 60, 67, 72, 77, 84 ], 4),
#         "duration": 0.5,
#         "amplitude": iso.PWhite(0, 128)
#     }, delay=1)
    amplitude = iso.PSequence([96,96,96,96], repeats=8)
    length = iso.PSequence([1/2,1/2,1/2,1/2], repeats=8)
    notes = iso.PSequence([1,3,2,4,3,5,4,6], repeats=4)
    timeline.schedule (
    {"note":notes+66,
     "duration":length/1,
     "amplitude":amplitude
    }
    )
    timeline.background()  # use background instead of run to enable live performing (async notes passing)
    
    amplitude = iso.PSequence([96,96,96,96], repeats=8)
    length = iso.PSequence([1/2,1/2,1/2,1/2], repeats=8)
    notes = iso.PSequence([1,3,2,4,3,5,4,6], repeats=4)
    timeline.schedule (
    {"note":notes+66,
     "duration":length/3,
     "amplitude":amplitude
    }, count=4
    )
    timeline.background()  # use background instead of run to enable live performing (async notes passing)
    
    
    
def play_test2():
  
    #--------------------------------------------------------------------------------
    # Play a series of 5 notes with random velocities.
    # Delay by 1 beat before playback.
    #--------------------------------------------------------------------------------
#     timeline.schedule({
#         "note": iso.PSequence([ 60, 67, 72, 77, 84 ], 4),
#         "duration": 0.5,
#         "amplitude": iso.PWhite(0, 128)
#     }, delay=1)
    amplitude = iso.PSequence([96,96,96,96], repeats=8)
    length = iso.PSequence([1/2,1/2,1/2,1/2], repeats=8)
    notes = iso.PSequence([1,3,2,4,3,5,4,6], repeats=4)
    timeline.schedule (
    {"note":notes+66,
     "duration":length/2,
     "amplitude":amplitude
    },delay=1
    )
    
    timeline.background()


tm_tst=play_test() 
# tm_tst=play_test2() 
print('Processing Done.')

[2022-05-21 22:56:25,680] Timeline: Scheduled new track (total tracks: 1)
[2022-05-21 22:56:25,686] Timeline: Starting
[2022-05-21 22:56:25,690] Timeline: Scheduled new track (total tracks: 2)
[2022-05-21 22:56:25,696] Timeline: Starting


Processing Done.


Exception in thread Thread-6:
Traceback (most recent call last):
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\threading.py", line 973, in _bootstrap_inner
    self.run()
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\timeline.py", line 245, in run
    raise e
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\timeline.py", line 234, in run
    self.clock_source.run()
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\clock.py", line 86, in run
    self.clock_target.tick()
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\timeline.py", line 149, in tick
    track.tick()
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\track.py", line 79, in tick
    self.note_offs.remove(note)
Va

 *** Exception in Timeline thread: list.remove(x): x not in list


[2022-05-21 22:56:26,001] Timeline: Track finished, removing from scheduler (total tracks: 1)
[2022-05-21 22:56:27,074] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:56:27,329] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:56:27,916] Clock: Timer overflowed (late by 0.046s)
[2022-05-21 22:56:29,403] Clock: Timer overflowed (late by 0.050s)
[2022-05-21 22:56:29,486] Clock: Timer overflowed (late by 0.023s)
[2022-05-21 22:56:29,722] Clock: Timer overflowed (late by 0.022s)
[2022-05-21 22:56:33,061] Clock: Timer overflowed (late by 0.022s)
[2022-05-21 22:56:33,679] Timeline: Track finished, removing from scheduler (total tracks: 0)
[2022-05-21 22:56:35,974] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:56:35,997] Clock: Timer overflowed (late by 0.021s)


In [4]:
play_test2()

[2022-05-21 22:29:03,222] Timeline: Starting
[2022-05-21 22:29:03,482] Timeline: Scheduled new track (total tracks: 2)
Exception in thread Thread-7:
Traceback (most recent call last):
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\threading.py", line 973, in _bootstrap_inner
    self.run()
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\timeline.py", line 245, in run
    raise e
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\timeline.py", line 234, in run
    self.clock_source.run()
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\clock.py", line 86, in run
    self.clock_target.tick()
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\timeline.py", line 149, in tick
    track.tick()
  File "C:\Users\piote\anac

 *** Exception in Timeline thread: list.remove(x): x not in list


[2022-05-21 22:29:04,270] Clock: Timer overflowed (late by 0.061s)
[2022-05-21 22:29:04,295] Clock: Timer overflowed (late by 0.047s)
[2022-05-21 22:29:06,239] Timeline: Track finished, removing from scheduler (total tracks: 1)
[2022-05-21 22:29:07,424] Timeline: Track finished, removing from scheduler (total tracks: 0)
[2022-05-21 22:29:07,637] Clock: Timer overflowed (late by 0.021s)
[2022-05-21 22:29:12,308] Clock: Timer overflowed (late by 0.035s)
[2022-05-21 22:29:12,527] Clock: Timer overflowed (late by 0.031s)
[2022-05-21 22:29:19,316] Clock: Timer overflowed (late by 0.025s)
[2022-05-21 22:29:21,791] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:29:22,083] Clock: Timer overflowed (late by 0.030s)
[2022-05-21 22:29:22,440] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:29:22,464] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:29:22,838] Clock: Timer overflowed (late by 0.027s)
[2022-05-21 22:29:22,949] Clock: Timer overflowed (late by 0.110s)
[2022-05

[2022-05-21 22:31:25,463] Clock: Timer overflowed (late by 0.028s)
[2022-05-21 22:31:26,362] Clock: Timer overflowed (late by 0.030s)
[2022-05-21 22:31:26,391] Clock: Timer overflowed (late by 0.027s)
[2022-05-21 22:31:26,932] Clock: Timer overflowed (late by 0.027s)
[2022-05-21 22:31:27,311] Clock: Timer overflowed (late by 0.030s)
[2022-05-21 22:31:30,064] Clock: Timer overflowed (late by 0.026s)
[2022-05-21 22:31:30,479] Clock: Timer overflowed (late by 0.026s)
[2022-05-21 22:31:33,827] Clock: Timer overflowed (late by 0.022s)
[2022-05-21 22:31:33,862] Clock: Timer overflowed (late by 0.034s)
[2022-05-21 22:31:33,901] Clock: Timer overflowed (late by 0.037s)
[2022-05-21 22:31:34,742] Clock: Timer overflowed (late by 0.029s)
[2022-05-21 22:31:35,957] Clock: Timer overflowed (late by 0.035s)
[2022-05-21 22:31:36,780] Clock: Timer overflowed (late by 0.032s)
[2022-05-21 22:31:38,067] Clock: Timer overflowed (late by 0.021s)
[2022-05-21 22:31:38,096] Clock: Timer overflowed (late by 0.0

[2022-05-21 22:34:32,937] Clock: Timer overflowed (late by 0.021s)
[2022-05-21 22:34:33,187] Clock: Timer overflowed (late by 0.023s)
[2022-05-21 22:34:33,687] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:34:33,737] Clock: Timer overflowed (late by 0.023s)
[2022-05-21 22:34:34,487] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:34:35,177] Clock: Timer overflowed (late by 0.023s)
[2022-05-21 22:34:36,072] Clock: Timer overflowed (late by 0.023s)
[2022-05-21 22:34:36,277] Clock: Timer overflowed (late by 0.023s)
[2022-05-21 22:34:36,424] Clock: Timer overflowed (late by 0.029s)
[2022-05-21 22:34:40,333] Clock: Timer overflowed (late by 0.022s)
[2022-05-21 22:34:41,131] Clock: Timer overflowed (late by 0.042s)
[2022-05-21 22:34:41,397] Clock: Timer overflowed (late by 0.021s)
[2022-05-21 22:34:42,032] Clock: Timer overflowed (late by 0.023s)
[2022-05-21 22:34:42,167] Clock: Timer overflowed (late by 0.024s)
[2022-05-21 22:34:42,939] Clock: Timer overflowed (late by 0.0

In [5]:
# timeline.schedule({
#     "action": lambda: print("Hello world")
# })
# timeline.schedule({
#     "action": lambda: print(round(timeline.current_time))
# }, count=4)

timeline.schedule({
   "action": lambda: print("Hello worldxx876x")
,
    "duration": 4
})

[2022-05-21 22:57:04,481] Timeline: Scheduled new track (total tracks: 2)


Hello worldxxx


Hello worldxx876x
Hello worldxxx
Hello worldxx876x
Hello worldxxx
Hello worldxx876x
Hello worldxxx
Hello worldxx876x
Hello worldxxx
Hello worldxx876x
Hello worldxxx
Hello worldxx876x


[2022-05-21 22:57:15,674] Clock: Timer overflowed (late by 0.021s)
[2022-05-21 22:57:15,697] Clock: Timer overflowed (late by 0.022s)


Hello worldxxx
Hello worldxx876x
Hello worldxxx
Hello worldxx876x


[2022-05-21 22:57:20,467] Clock: Timer overflowed (late by 0.045s)


Hello worldxxx
Hello worldxx876x


In [22]:
def set_key(k):
    iso.Globals.set("key", iso.Key(key))

timeline.schedule({
    "action": set_key,
    "args": {
        "k": iso.PWhite(8)
    },
    "duration": 4
})
timeline.schedule({
    "degree": 0,
    "key": iso.PGlobals("key"),
    "octave": 4
})

[2022-05-21 22:54:41,668] Timeline: Scheduled new track (total tracks: 2)
[2022-05-21 22:54:41,670] Timeline: Scheduled new track (total tracks: 3)


Traceback (most recent call last):
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\track.py", line 219, in perform_event
    event.action(**event.args)
  File "C:\Users\piote\AppData\Local\Temp\ipykernel_12000\2537458868.py", line 2, in set_key
    iso.Globals.set("key", iso.Key(key))
NameError: name 'key' is not defined
Exception in thread Thread-27:
Traceback (most recent call last):
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\threading.py", line 973, in _bootstrap_inner
    self.run()
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\timeline.py", line 245, in run
    raise e
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-packages\isobar\timeline\timeline.py", line 234, in run
    self.clock_source.run()
  File "C:\Users\piote\anaconda3\envs\soundDesign\lib\site-

Exception when handling scheduled action: name 'key' is not defined
 *** Exception in Timeline thread: Global variable does not exist: key
